insights/parsers/ceph_insights.py

```diff
@@ -0,0 +1,93 @@
+"""
+ceph_insights - command ``ceph insights``
+=========================================
+"""
+import json
+import re
+from .. import CommandParser, parser
+from insights.specs import Specs
+
+
+@parser(Specs.ceph_insights)
+class CephInsights(CommandParser):
+    """
+    Parse the output of the ``ceph insights`` command.
+
+    Attributes:
+        version (dict): version information of the Ceph cluster.
+        data (dict): a dictionary of the parsed output.
+
+    The ``data`` attribute is a dictionary containing the parsed output of the
+    ``ceph insights`` command. The following are available in ``data``:
+
+    * ``crashes`` - summary of daemon crashes for the past 24 hours
+    * ``health`` - the current and historical (past 24 hours) health checks
+    * ``config`` - cluster and daemon configuration settings
+    * ``osd_dump`` - osd and pool information
+    * ``df`` - storage usage statistics
+    * ``osd_tree`` - osd topology
+    * ``fs_map`` - file system map
+    * ``crush_map`` - the CRUSH map
+    * ``mon_map`` - monitor map
+    * ``service_map`` - service map
+    * ``manager_map`` - manager map
+    * ``mon_status`` - monitor status
+    * ``pg_summary`` - placement group summary
+    * ``osd_metadata`` - per-OSD metadata
+    * ``version`` - ceph software version
+    * ``errors`` - any errors encountered collecting this data
+
+    The ``version`` attribute contains a normalized view of ``self.data["version"]``.
+
+    Examples:
+        >>> ceph_insights.version["release"] == 14
+        True
+        >>> ceph_insights.version["major"] == 0
+        True
+        >>> ceph_insights.version["minor"] == 0
+        True
+        >>> isinstance(ceph_insights.data["crashes"], dict)
+        True
+        >>> isinstance(ceph_insights.data["health"], dict)
+        True
+    """
+    IGNORE_RE = [
+         "\*\*\* DEVELOPER MODE",
+         "\d+-\d+-\d+.+WARNING: all dangerous"
+    ]
+
+    bad_lines = CommandParser.bad_lines + [
+        "module 'insights' is not enabled",
+        "no valid command found"
+    ]
```

Comment:
@noahdesu this is the first time anyone has extended `bad_lines` in this manner.  Rather that permanently tie the internal implementation to a list of bad lines, would you mind to implement a new class method `add_bad_lines` in the `CommandParser` class that can be invoked in the sub class body (replacing this assignment) to add a list of subclass lines to the CommandParser?

Reply:
@noahdesu Also want to rename `CommandParser.bad_lines` to `CommandParser.__bad_lines` so subclasses don't accidentally shadow it.

Reply:
@csams @bfahr do you want this to be a class variable, rather than an instance variable? in the former case changes to bad_lines will be seen by all other command parser instance. caveat: perhaps there is a python trick / pattern I'm not aware of?

Reply:
@noahdesu actually, the `CommandParser`'s logic should be for lines that might indicate failure for *any* command, not just a particular one. Since the lines in this case are specific to the command, the parser itself should just check for them in its `parse_content` function in a way that makes sense to it. We should still adjust `CommandParser.bad_lines` to `CommandParser.__bad_lines` so subclasses don't accidentally clobber it, but I don't think we should provide a general API for this. The `CommandParser` exists only to prevent every parser from needing to deal with a collection bug in the wild that mistakes error messages from the shell as output from commands.

Reply:
I'm fine with moving the check into the Ceph parser, but the semantics are exactly the same except for the strings being search for, so it'd be nice to reuse that. Another option would be to add an optional argument to CommandParser constructor. Something like:

```python
  class CommandParser(Parser):
    def __init__(self, context, extra_bad_lines=[]):
      self.validate_lines(context.content, self.__badlines + extra_bad_lines)

    @staticmethod
    def validate_lines(results, bad_lines):
       ....
```

Reply:
@noahdesu I didn't want to make it part of the API for subclasses since we want to remove the check from the base class when the old collectors with the bug are deprecated, but we can remove the check in a smarter way while allowing subclasses to reuse the validation code.

Go ahead with the constructor keyword. Just protect the concatenation so we're not creating a new list on every call if `extra_bad_lines` is empty.

Does the ceph command have a nonzero return code in the cases you're protecting against?

Reply:
pushed changes for this. if it looks ok i'll let it sit for now until #1455 is done.